psybot/modules/challenge.py

```diff
@@ -107,47 +112,77 @@ async def add(interaction: discord.Interaction, category: str, name: str):
+    category = sanitize_channel_name(category) if category else None
+    settings = get_settings(interaction.guild)
+    if settings.enforce_categories and not category_is_valid(category, interaction.guild_id):
+        raise app_commands.AppCommandError("Invalid CTF category")
+
     ctf = sanitize_channel_name(ctf_db.name)
-    name = sanitize_channel_name(name)
 
-    if category:
-        category = sanitize_channel_name(category)
-        fullname = f"{ctf}-{category}-{name}"
-    else:
-        category = None
-        fullname = f"{ctf}-{name}"
+    class ChallengeInfoModal(ui.Modal, title="Add Challenge"):
+        name_field = ui.TextInput(
+            label="Name",
+            default=name,
+            placeholder="Challenge name"
+        )
+        category_field = ui.TextInput(
+            label="Category",
+            default=category,
+            placeholder="Challenge category"
+        )
+        description_field = ui.TextInput(
+            label="Description",
+            style=discord.TextStyle.paragraph,
+            placeholder="Challenge description",
+            max_length=1000
+        )
+
+        async def on_submit(self, submit_interaction: discord.Interaction):
+            name = sanitize_channel_name(self.name_field.value)
+            if self.category_field.value:
+                category = sanitize_channel_name(self.category_field.value)
+                if settings.enforce_categories and not category_is_valid(category, interaction.guild_id):
+                    await submit_interaction.response.send_message("Invalid CTF category", ephemeral=True)
+                    return
+                channel_name = f"{ctf}-{category}-{name}"[:100]
+            else:
+                category = None
+                channel_name = f"{ctf}-{name}"[:100]
 
-    settings = get_settings(interaction.guild)
-    if settings.enforce_categories:
-        if category is not None and not CtfCategory.objects(name=category, guild_id=interaction.guild_id):
-            raise app_commands.AppCommandError("Invalid CTF category")
+            if old_chall := Challenge.objects(name=name, category=category, ctf=ctf_db).first():
+                if interaction.guild.get_channel(old_chall.channel_id):
+                    await submit_interaction.response.send_message("A challenge with that name already exists", ephemeral=True)
+                    return
+                else:
+                    old_chall.delete()
 
-    if old_chall := Challenge.objects(name=name, category=category, ctf=ctf_db).first():
-        if interaction.guild.get_channel(old_chall.channel_id):
-            raise app_commands.AppCommandError("A challenge with that name already exists")
-        else:
-            old_chall.delete()
+            new_channel = await create_channel(channel_name, interaction.channel.overwrites, incomplete_category)
+            await new_channel.send(f"# {self.name_field.value}\n\n{self.description_field.value}")
```

Comment:
I don't think this will be an issue, but we might run into very long descriptions.

This can at max be 2000 characters for a bot, and I don't see any character limit on the modal (I suspect each input can be up to 2000 characters, but I don't know for sure.

So, this might fail.

Reply:
Modal description field actually has a `max_length` set to 1000 right now, so it should be okay.

Reply:
Amazing. I couldn't find any documentation on it, but that would keep us within limits.